File: go-key-service/internal/signing/signing.go
// Package signing produces Ed25519 signatures over canonical action bytes.
// The Go and Python implementations MUST agree byte-for-byte; see
// docs/schema.md and docs/signing_vectors.json.
package signing

import (
	"crypto/ed25519"
	"errors"
	"fmt"
	"io"

	"github.com/knhn1004/CryptoAgent/go-key-service/internal/action"
)

var (
	ErrInvalidSignature = errors.New("signing: invalid signature")
	ErrInvalidKeyLength = errors.New("signing: invalid key length")
)

// Sign canonicalizes the action and returns its Ed25519 signature.
func Sign(a *action.Action, priv ed25519.PrivateKey) ([]byte, error) {
	if len(priv) != ed25519.PrivateKeySize {
		return nil, fmt.Errorf("%w: private key %d bytes", ErrInvalidKeyLength, len(priv))
	}
	msg, err := a.Canonical()
	if err != nil {
		return nil, err
	}
	return ed25519.Sign(priv, msg), nil
}

// Verify returns nil iff sig is a valid Ed25519 signature on the canonical
// bytes of a under pub.
func Verify(a *action.Action, sig []byte, pub ed25519.PublicKey) error {
	if len(pub) != ed25519.PublicKeySize {
		return fmt.Errorf("%w: public key %d bytes", ErrInvalidKeyLength, len(pub))
	}
	msg, err := a.Canonical()
	if err != nil {
		return err
	}
	if !ed25519.Verify(pub, msg, sig) {
		return ErrInvalidSignature
	}
	return nil
}

// GenerateKey wraps ed25519.GenerateKey so callers don't need to import
// crypto/ed25519 directly.
func GenerateKey(rand io.Reader) (ed25519.PublicKey, ed25519.PrivateKey, error) {
	return ed25519.GenerateKey(rand)
}

File: sdk-python/cryptoagent/signing.py
"""Ed25519 sign/verify over the canonical action schema.

Mirror of go-key-service/internal/signing/signing.go. Both libraries MUST
produce byte-identical signatures for equal (seed, action) inputs;
docs/signing_vectors.json is the authoritative cross-language fixture.

Keys here use the 32-byte raw Ed25519 seed (== first 32 bytes of Go's
ed25519.PrivateKey, which stores seed||public_key).
"""

from __future__ import annotations

import os

from nacl.exceptions import BadSignatureError
from nacl.signing import SigningKey, VerifyKey

from .action import Action

PUBLIC_KEY_LEN = 32
PRIVATE_KEY_SEED_LEN = 32
SIGNATURE_LEN = 64


class SignatureError(Exception):
    """Verification failed."""


def generate_keypair() -> tuple[bytes, bytes]:
    """Returns ``(public_key_32, private_key_seed_32)``."""
    seed = os.urandom(PRIVATE_KEY_SEED_LEN)
    sk = SigningKey(seed)
    return bytes(sk.verify_key), seed


def sign(action: Action, private_key: bytes) -> bytes:
    if len(private_key) != PRIVATE_KEY_SEED_LEN:
        raise ValueError(
            f"private key must be {PRIVATE_KEY_SEED_LEN}-byte seed, got {len(private_key)}"
        )
    sk = SigningKey(private_key)
    msg = action.canonical()
    return bytes(sk.sign(msg).signature)


def verify(action: Action, signature: bytes, public_key: bytes) -> None:
    if len(public_key) != PUBLIC_KEY_LEN:
        raise ValueError(f"public key must be {PUBLIC_KEY_LEN} bytes, got {len(public_key)}")
    if len(signature) != SIGNATURE_LEN:
        raise SignatureError(f"signature must be {SIGNATURE_LEN} bytes, got {len(signature)}")
    vk = VerifyKey(public_key)
    msg = action.canonical()
    try:
        vk.verify(msg, signature)
    except BadSignatureError as e:
        raise SignatureError("invalid signature") from e
